beehive-lab/TornadoVM
# [FPGA] Enhancement of Xilinx OpenCL backend

#### Description

This PR provides enhancements for the generated code that target Xilinx FPGAs.

In particular two features are introduced:
- The ability to instantiate multiple compute units for the same kernel.
- The transparent addition of the `__attribute__((xcl_pipeline_loop(1)))`. (1 is the default initiation interval value (II) used by Xilinx HLS tools).

#### Problem description

1. With regards to the first feature, we have seen up to 2.5x performance improvement when fitting multiple compute units on the FPGA. In particular for the DFT and 65536 input size, we managed to fit up to 12 kernels on the Xilinx KCU1500 FPGA. The problem that did not allow this flag to work was due to an inability in the `OCLCodeCache` to parse the `--nk computeDft:2` as a flag in the FPGA configuration file.
This flag is tested with SDAccel compiler `xocc` and resulted in 2.5x speedup for `DFTDynamic`.

2. Regarding the second feature, TornadoVM currently supports the loop unrolling for Intel FPGAs. This PR has added the addition of the Xilinx OpenCL pipelining attribute for loops, as this attribute can contribute to up to 2.2x performance increase. The XIlinx OpenCL loop unrolling attribute did not result in any performance change.

#### Backend/s tested


#### OS tested


#### Did you check on FPGAs?

If possible, check your changes on FPGAs.


#### How to test the new patch?

I tested the PR in the Intel FPGAs to ensure that it does not break the code generation. In addition, I tested it on the Xilinx KCU1500 FPGA using SDAccel 2018.2 and on emulation mode with Vitis 2020.2. Here is the output of the execution with the generated code:

```bash
$ tornado --printKernel --debug --igv -Ds0.t0.device=0:1 uk.ac.manchester.tornado.examples.dynamic.DFTDynamic 512 tornado 1

__kernel void lookupBufferAddress(__global uchar *_heap_base, ulong _frame_base, __constant uchar *_constant_region, __local uchar *_local_region, __global int *_atomics)
{

  __global ulong *_frame = (__global ulong *) &_heap_base[_frame_base];


  // BLOCK 0
  _frame[0]  =  (ulong) _heap_base;
}  //  kernel

Initialization time:  759578609 ns

__kernel void lookupBufferAddress(__global uchar *_heap_base, ulong _frame_base, __constant uchar *_constant_region, __local uchar *_local_region, __global int *_atomics)
{

  __global ulong *_frame = (__global ulong *) &_heap_base[_frame_base];


  // BLOCK 0
  _frame[0]  =  (ulong) _heap_base;
}  //  kernel

[DEBUG] JIT compilation for the FPGA
Warning: -Dtornado.opencl.userelative was set to False. TornadoVM changed it to True because it is required for FPGA execution.
__attribute__((reqd_work_group_size(64,1,1)))
__kernel void computeDft(__global uchar *_heap_base, ulong _frame_base, __constant uchar *_constant_region, __local uchar *_local_region, __global int *_atomics)
{
  long l_11, l_30, l_32, l_31, l_10, l_9; 
  float f_28, f_27, f_26, f_25, f_24, f_7, f_6, f_15, f_13, f_23, f_22, f_21, f_20, f_19, f_18, f_17, f_16; 
  ulong ul_12, ul_14, ul_0, ul_2, ul_34, ul_1, ul_33, ul_3; 
  int i_8, i_29, i_35, i_5, i_4, i_36; 

  __global ulong *_frame = (__global ulong *) &_heap_base[_frame_base];


  // BLOCK 0
  ul_0  =  (ulong) _frame[3];
  ul_1  =  (ulong) _frame[4];
  ul_2  =  (ulong) _frame[5];
  ul_3  =  (ulong) _frame[6];
  i_4  =  get_global_id(0);
  // BLOCK 1 MERGES [0 5 ]
  i_5  =  i_4;
  // BLOCK 2
  // BLOCK 3 MERGES [2 4 ]
  f_6  =  0.0F;
  f_7  =  0.0F;
  i_8  =  0;
  __attribute__((xcl_pipeline_loop(1)))
  for(;i_8 < 512;)
  {
    // BLOCK 4
    l_9  =  (long) i_8;
    l_10  =  l_9 << 2;
    l_11  =  l_10 + 24L;
    ul_12  =  ul_0 + l_11;
    f_13  =  *((__global float *) &_heap_base[ul_12]);
    ul_14  =  ul_1 + l_11;
    f_15  =  *((__global float *) &_heap_base[ul_14]);
    f_16  =  (float) i_8;
    f_17  =  f_16 * 6.2831855F;
    f_18  =  (float) i_5;
    f_19  =  f_17 * f_18;
    f_20  =  f_19 / 512.0F;
    f_21  =  sin(f_20);
    f_22  =  cos(f_20);
    f_23  =  f_22 * f_15;
    f_24  =  fma(f_21, f_13, f_23);
    f_25  =  f_7 - f_24;
    f_26  =  f_21 * f_15;
    f_27  =  fma(f_22, f_13, f_26);
    f_28  =  f_6 + f_27;
    i_29  =  i_8 + 1;
    f_6  =  f_28;
    f_7  =  f_25;
    i_8  =  i_29;
  }  // B4
  // BLOCK 5
  l_30  =  (long) i_5;
  l_31  =  l_30 << 2;
  l_32  =  l_31 + 24L;
  ul_33  =  ul_2 + l_32;
  *((__global float *) &_heap_base[ul_33])  =  f_6;
  ul_34  =  ul_3 + l_32;
  *((__global float *) &_heap_base[ul_34])  =  f_7;
  i_35  =  get_global_size(0);
  i_36  =  i_35 + i_5;
  i_5  =  i_36;
  // BLOCK 6
  return;
}  //  kernel

Task info: s0.t0
	Backend           : OpenCL
	Device            : xilinx_u50_gen3x16_xdma_201920_3 CL_DEVICE_TYPE_ACCELERATOR (available)
	Dims              : 1
	Global work offset: [0]
	Global work size  : [512]
	Local  work size  : [64, 1, 1]
	Number of workgroups  : [8]

Total time:  58947809342 ns 

Is valid?: true

Validation: SUCCESS 
```

Provide instructions about how to test the new patch.
